more_itertools/more.py

```diff
@@ -1898,3 +1901,68 @@ def inner_wrapper(*args, **kwargs):
         return outer_wrapper
 
     return decorator
+
+
+class classify(object):
+    """Wrap *iterable* and return an object that progressively saves its items
+    based on a *key* function.
+
+        >>> iterable = ['a1', 'b1', 'c1', 'a2', 'b2', 'c2', 'b3']
+        >>> key = lambda x: x[0]  # Classify by first character
+        >>> it = classify(iterable, key=key)
+        >>> next(it), next(it), next(it)  # Advance the iterable
+        ('a1', 'b1', 'c1')
+        >>> it['a']  # 'a' items seen so far
+        ['a1']
+        >>> list(it)  # Exhaust the iterable
+        ['a2', 'b2', 'c2', 'b3']
+        >>> it['a']
+        ['a1', 'a2']
+        >>> it['b']
+        ['b1', 'b2', 'b3']
+
+    :class:`classify` automates collecting items into a ``defaultdict(list)``
+    object:
+
+        >>> from collections import defaultdict
+        >>> mapping = defaultdict(list)
+        >>> for item in iterable:
+        ...     category = key(item)
+        ...     mapping[category].append(item)
+
+    Access the underlying :obj:`defaultdict` through the :attr:`mapping`
+    attribute:
+
+        >>> 'a' in it.mapping
```

Comment:
I suppose we oughtn't give this as an example, as it looks like you meant it to be more idiomatic to say `'a' in it` itself, given `__contains__` below.